R/conformal_infer_cv.R

```diff
@@ -0,0 +1,280 @@
+  invisible(x)
+}
+
+# ------------------------------------------------------------------------------
+# helpers
+
+new_infer_cv <- function(models, resid) {
+  if (!is.numeric(resid)) {
+    rlang::abort("Absolute residuals should be numeric")
+  }
+  na_resid <- is.na(resid)
+  if (all(na_resid)) {
+    rlang::abort("All of the absolute residuals are missing.")
+  }
+
+  if (!is.list(models)) {
+    rlang::abort("The model list should be... a list")
+  }
+  is_wflow <- purrr::map_lgl(models, workflows::is_trained_workflow)
+  if (all(!is_wflow)) {
+    rlang::abort(".extracts does not contain fitted workflows")
+  }
+  if (any(!is_wflow)) {
+    models <- models[is_wflow]
+  }
+
+  res <- list(
+    models = models,
+    abs_resid = resid[!na_resid]
+  )
+  class(res) <- "int_conformal_infer_cv"
+  res
+}
+
+.get_lower_cv_bound <- function(pred, resid, level = 0.95) {
+  as.vector(stats::quantile(pred - resid, probs = 1 - level))
+}
+
+.get_upper_cv_bound <- function(pred, resid, level = 0.95) {
+  as.vector(stats::quantile(pred + resid, probs = level))
+}
+
+.get_pre_type <- function(x) {
+  cls <- x %>% workflows::extract_preprocessor() %>% class()
+  cls <- cls[!grepl("butchered", cls)]
+  cls[1]
+}
+
+.get_fit_type <- function(x) {
+  fitted <- x %>% workflows::extract_fit_parsnip()
+  res <- paste0(class(fitted$spec)[1], " (engine = ", fitted$spec$engine, ")")
+  res
+}
+
+.get_fitted_workflows <- function(x, prm = NULL) {
+  if (is.null(prm)) {
+    res <- purrr::map(x$.extracts, ~ .x$.extracts[[1]])
+  } else {
+    by_vars <- names(prm)
+    res <-
+      x %>%
+      dplyr::select(.extracts) %>%
+      tidyr::unnest(.extracts) %>%
+      dplyr::inner_join(prm, by = by_vars) %>%
+      purrr::pluck(".extracts")
+  }
+  res
+}
+
+# ------------------------------------------------------------------------------
+# checks
+
+check_resampling <- function(x) {
+  rs <- attr(x, "rset_info")
+  if (rs$att$class != "vfold_cv") {
+    msg <- paste0(
+      "The data were resampled using ", rs$label,
+      ". This method was developed for V-fold cross-validation. Interval ",
+      "coverage is unknown for your resampling method."
+    )
+    rlang::warn(msg)
+  } else {
+    if (rs$att$repeats > 1) {
+      msg <- paste0(
+        rs$att$repeats, " repeats were used. This method was developed for ",
+        "basic V-fold cross-validation. Interval coverage is unknown multiple ",
+        "repeats."
+      )
+      rlang::warn(msg)
+    }
+  }
+  invisible(NULL)
+}
+
+check_parameters <- function(x, param) {
+  prms <- tune::.get_tune_parameter_names(x)
+  mtr <- tune::collect_metrics(x) %>%
+    dplyr::distinct(.config, !!!rlang::syms(prms))
+  remain <- dplyr::inner_join(mtr, param, by = names(param))
+  if (nrow(remain) > 1) {
+    msg <-
+      paste0(
+        "The `parameters` argument selected ", nrow(remain), " submodels. Only ",
+        "1 should be selected."
+      )
+    rlang::abort(msg)
```

Comment:
Could we pass `call` in these `abort()` calls and test with `expect_snapshot(error = TRUE, ...)` rather than `expect_snapshot_error(...)` to test the error context?